### ISSUE TITLE
Better support for map printing

### ISSUE THREAD

**Issue by tyroguru** [bpftrace/bpftrace#2952]
A user should be able to specify precise formatting for the tuple entries in the map (the aggregation) and also the aggregating function result as well. So, if we have a very simple example of a one element tuple with a count() function, e.g. if we were tracking pointers, then we should be able to specify that the tuple member is a pointer when we print it. We can do that with tuples and `printf()` but not with `print()`. For example:

```
# cat print-formatting.bt
BEGIN
{
  @[0xfeedface] = count();
  @[0xbaadcafe] = count();

  @x = (0xfeedface, 0xbaadcafe);
}

interval:s:1
{
  print(@);
  printf("%p %p\n", @x.0, @x.1);
}

# sudo bpftrace ./print-formatting.bt
Attaching 2 probes...
@[4277009102]: 1
@[3131951870]: 1
0xfeedface 0xbaadcafe
```

Using the simple example above we should, at a minimum be able to do the following to specify the first field of the aggregation should be printed as a pointer.:

```
print(“%p”, @);
```

Just to extend it further clarification,  if the tuple was @[void*, int, char] I should be able to do:

```
print(“%p %d %c”, @);
```

We should also be able to control the format modifier for the aggregating function result as well.

**Comment by ajor** [comment]
I'd prefer that we keep explicit formatting out of the `print()` function, but there are two other enhancements (both useful by themselves) that we could make to improve this type of map printing.

### Approach 1
Allow declaring map types up front: #2954
```
let @[pointer_t] = count_t; // syntax TBD
BEGIN
{
  @[0xfeedface] = count();
  @[0xbaadcafe] = count();
  print(@);
}
```
Since we know that the map's key has a pointer type, `print()` could automatically print it as a pointer.

### Approach 2
Provide better looping constructs to allow scripts to iterate over the keys of a map and print them out individually:
```
@[0xfeedface] = count();
@[0xbaadcafe] = count();
for ($key : @)
{
  printf("@[%p]: %d\n", $key, @[$key]);
}
```

**Comment by jordalgo** [comment]
Just detailing some inconsistencies:
```
sudo bpftrace -e 'BEGIN { @ = count(); print((@)); exit(); }'
Attaching 1 probe...
@: 1

@: 1
```

```
sudo bpftrace -e 'BEGIN { @ = count(); print((@, 1)); exit(); }'
Attaching 1 probe...
(1, 1)

@: 1
```

```
sudo bpftrace -e 'BEGIN { @[1] = 2; @[2] = 3; print((@)); exit(); }'
Attaching 1 probe...
@[1]: 2
@[2]: 3

@[1]: 2
@[2]: 3
```

```
sudo bpftrace -e 'BEGIN { @[1] = 2; @[2] = 3; print((@, 1)); exit(); }'
stdin:1:29-37: ERROR: Argument mismatch for @: trying to access with arguments: [] when map expects arguments: [unsigned int64]
BEGIN { @[1] = 2; @[2] = 3; print((@, 1)); exit(); }
                            ~~~~~~~~
```

**Comment by ajor** [comment]
Closing as resolved by #3003.

Something similar to the original example can now be implemented as:
```
$ sudo bpftrace -e '
    BEGIN
    {
      @[0xfeedface] = 1;
      @[0xbaadcafe] = 2;
      for ($kv : @)
      {
        printf("@[%p]: %d\n", $kv.0, $kv.1);
      }
    }'
Attaching 1 probe...
@[0xfeedface]: 1
@[0xbaadcafe]: 2
^C
```
